src/spox/opset/ai/onnx/v21.py

```diff
@@ -612,6 +612,20 @@ class Inputs(BaseInputs):
     class Outputs(BaseOutputs):
         v_final_and_scan_outputs: Sequence[Var]
 
+    def infer_output_types(self) -> Dict[str, Type]:
+        output_types = super().infer_output_types()
+
+        body = self.attrs.body.value
+        n = len(body.requested_arguments) - 2
+
+        carried_names = list(self.outputs.get_vars())[:n]
+        carried_types = [v.type for v in list(body.requested_results.values())[1:][:n]]
+
+        for name, typ in zip(carried_names, carried_types):
+            output_types[name] = typ
+
```

Comment:
Did you not want to check the output type of the body aligns with the initial value as per https://github.com/Quantco/spox/issues/163#issuecomment-2242788733?

Reply:
I'm not using loops in the end, so that's fine.

But I think this only make sense if you can explicitly mark dimensions that you want to change in the loop as `None`.

Let's say I want to use a function `lambda x: concat([x, [0]], axis=0)` in the loop.
If my input has shape `(None,)`, then the shape inference passes, because the output is also `(None,)`.
However, if my input has known shape, it will fail.

So it's not just very limiting, it's also inconsistent.

These issues should go away if you can say "This dimensions will be changed, treat it as `None`". If the number of dimensions will also change, you set the whole shape to `None`.